samples/csharp/getting-started/Ranking_AutoML/Ranking/Ranking/Program.cs

```diff
@@ -0,0 +1,232 @@
+            var firstGroupId = searchQueries.First().GroupId;
+            var firstGroupPredictions = searchQueries.Take(100).Where(p => p.GroupId == firstGroupId).OrderByDescending(p => p.Score).ToList();
+
+            // The individual scores themselves are NOT a useful measure of result quality; instead, they are only useful as a relative measure to other scores in the group. 
+            // The scores are used to determine the ranking where a higher score indicates a higher ranking versus another candidate result.
+            foreach (var prediction in firstGroupPredictions)
+            {
+                Console.WriteLine($"GroupId: {prediction.GroupId}, Score: {prediction.Score}");
+            }
+        }
+
+        private static void GetData(string inputPath, string outputPath, string trainDatasetPath, string trainDatasetUrl,
+            string testDatasetUrl, string testDatasetPath, string validationDatasetUrl, string validationDatasetPath)
+        {
+            Console.WriteLine("===== Prepare data =====\n");
+
+            if (!Directory.Exists(outputPath))
+            {
+                Directory.CreateDirectory(outputPath);
+            }
+
+            if (!Directory.Exists(inputPath))
+            {
+                Directory.CreateDirectory(inputPath);
+            }
+
+            if (!File.Exists(trainDatasetPath))
+            {
+                Console.WriteLine("===== Downloading the train dataset - this may take several minutes =====\n");
+                using (var client = new WebClient())
+                {
+                    client.DownloadFile(trainDatasetUrl, TrainDatasetPath);
+                }
+            }
+
+            if (!File.Exists(validationDatasetPath))
+            {
+                Console.WriteLine("===== Downloading the validation dataset - this may take several minutes =====\n");
+                using (var client = new WebClient())
+                {
+                    client.DownloadFile(validationDatasetUrl, validationDatasetPath);
+                }
+            }
+
+            if (!File.Exists(testDatasetPath))
+            {
+                Console.WriteLine("===== Downloading the test dataset - this may take several minutes =====\n");
+                using (var client = new WebClient())
+                {
+                    client.DownloadFile(testDatasetUrl, testDatasetPath);
+                }
+            }
+
+            Console.WriteLine("===== Download is finished =====\n");
+        }
+
+        private static void PrintTopModels(ExperimentResult<RankingMetrics> experimentResult)
+        {
+            // Get top few runs ranked by nDCG
+            var topRuns = experimentResult.RunDetails
+                .Where(r => r.ValidationMetrics != null && !double.IsNaN(r.ValidationMetrics.NormalizedDiscountedCumulativeGains[0]))
+                .OrderByDescending(r => r.ValidationMetrics.NormalizedDiscountedCumulativeGains[0]).Take(3);
```

Comment:
Looking at the AutoML code, it defaults to NDCG@10 (or the last element of the array):
https://github.com/dotnet/machinelearning/blob/f87a3bbd8adb12934dae0a0060813ce9b7500664/src/Microsoft.ML.AutoML/Experiment/MetricsAgents/RankingMetricsAgent.cs#L31-L32

Should use the same sorting logic here. This will ensure that what the AutoML code returns as its `experimentResult.BestRun` will show sorted to the top of this list.

---

As an aside the AutoML code isn't passing through a `k` parameter for the optimization metric. This is room for improvement in our AutoML ranking API/code. Currently the AutoML code is hard-coded to optimize NDCG@10 (or DCG@10) instead of a user supplied `k` in `NDCG@k`. Having the user supply this parameter instead of hard coding lets the AutoML code optimize the model towards the user's needs; for instance NDCG@3 is useful for an auto-suggest showing three results, NDCG@10 is useful for the first page of search results, NDCG@30 optimizes the first three pages.

Reply:
I'll put an issue for this today to help keep track of this. I may even try to implement it 😄

Reply:
Setting to sort on NDCG@10 to match how [AutoML selects the best model](https://github.com/dotnet/machinelearning/blob/f87a3bbd8adb12934dae0a0060813ce9b7500664/src/Microsoft.ML.AutoML/Experiment/MetricsAgents/RankingMetricsAgent.cs#L31-L32):
```suggestion
                .OrderByDescending(r => r.ValidationMetrics.NormalizedDiscountedCumulativeGains[9]).Take(3);
```

We could still match the logic a bit closer (by duplicating the code to here).

Reply:
#Resolved